ranges/stackarrays.nim

```diff
@@ -0,0 +1,57 @@
+proc alloca(n: int): pointer {.importc, header: "<alloca.h>".}
+
+type
+  StackArray*[T] = ptr object
+    bufferLen: int
+    buffer: UncheckedArray[T]
+
+template `[]`*(a: StackArray, i: int): auto =
+  if i < 0 or i >= a.len: raise newException(RangeError, "index out of range")
+  a.buffer[i]
+
+proc `[]=`*(a: StackArray, i: int, val: a.T) =
+  if i < 0 or i >= a.len: raise newException(RangeError, "index out of range")
+  a.buffer[i] = val
+
+proc len*(a: StackArray): int {.inline.} =
+  a.bufferLen
+
+template high*(a: StackArray): int =
+  a.bufferLen - 1
+
+template low*(a: StackArray): int =
+  0
+
+iterator items*(a: StackArray): a.T =
+  for i in 0 .. a.high:
+    yield a.buffer[i]
+
+iterator mitems*(a: var StackArray): var a.T =
+  for i in 0 .. a.high:
+    yield a.buffer[i]
+
+iterator pairs*(a: StackArray): a.T =
+  for i in 0 .. a.high:
+    yield (i, a.buffer[i])
+
+iterator mpairs*(a: var StackArray): (int, var a.T) =
+  for i in 0 .. a.high:
+    yield (i, a.buffer[i])
+
+template allocStackArray*(T: typedesc, size: int): auto =
+  if size < 0: raise newException(RangeError, "allocation with a negative size")
+  # XXX: is it possible to perform a stack size check before calling `alloca`?
+  # On thread init, Nim may record the base address and the capacity of the stack,
+  # so in theory we can verify that we still have enough room for the allocation.
+  # Research this.
+  var
+    bufferSize = size * sizeof(T)
+    totalSize = sizeof(int) + bufferSize
+    arr = cast[StackArray[T]](alloca(totalSize))
+  zeroMem(addr arr.buffer[0], bufferSize)
+  arr.bufferLen = size
+  arr
+
+template toOpenArray*(a: StackArray): auto =
```

Comment:
This procedure converts whole array, but i think it would be better to have template which supports slices too.